src/loader/bgw_counter.c

```diff
@@ -122,11 +122,16 @@ bgw_total_workers_decrement()
 	 */
 	SpinLockAcquire(&ct->mutex);
 	if (ct->total_workers > 1)
+	{
 		ct->total_workers--;
+		SpinLockRelease(&ct->mutex);
+	}
 	else
+	{
+		SpinLockRelease(&ct->mutex);
```

Comment:
FWIW, I think as a rule-of-thumb, we shouldn't throw errors at all in these types of functions (lower-level utility-functions, state-change functions, etc.). This has been a problem also elsewhere (e.g., networking). The issue is that there is no way to know, from higher-level code, that some low-level deeply nested function will throw an error since there is no way to signal exceptions being thrown. That makes it easy to make mistakes w.r.t. cleaning up state, including locks. The only valid exceptions to throw in code like this should be `FATAL` for unrecoverable errors, basically aborting the process.

Returning an error properly signals that there are error cases that should be handled. I think we should reserve throwing errors to high-level functions, or, we need to wrap calls in PG_TRY() clauses.

IMO, we should return proper errors and handle them. If we need to bail out and report errors, it should be done as high-up in the calling stack as possible.

Reply:
I'm going to change this to a log as it will affect whether we are utilizing resources correctly, but probably should not result in an error.